### PR TITLE
Disable AI search when literal search finds no matches

### PR DESCRIPTION
**Pull request by Mbeaulne**
## Description

AI search is now disabled when a literal (lexical) search returns no matches. Previously, when no lexical hits were found, AI search would fall back to a sampled "browse pool" built by evenly sampling the full component index, allowing the AI reranker to run against a representative spread of components. This behavior has been removed — if the query doesn't match anything lexically, the AI search button is disabled and the reranker is not called.

This removes the `sampleEvenly` utility function and the `indexEntryToLexicalMatch` fallback path from the Dashboard view, and updates `buildAiCandidateMatches` to return an empty array immediately when lexical search finds nothing.

## Type of Change

- [ ] Bug fix
- [ ] New feature
- [x] Improvement
- [x] Cleanup/Refactor
- [ ] Breaking change
- [ ] Documentation update

## Checklist

- [ ] I have tested this does not break current pipelines / runs functionality
- [ ] I have tested the changes on staging

## Test Instructions

1. Open the Dashboard components view with AI search configured.
2. Enter a query that has no lexical matches (e.g., a nonsense string like `qqzznomatch`).
3. Confirm the "AI search" button is disabled and clicking it does not trigger a rerank call.
4. Enter a query that does have lexical matches and confirm AI search remains functional.

## Additional Comments

The previous browse-pool fallback was intended to give the AI model a representative sample when literal search found nothing. This PR removes that behavior, meaning AI search is only available as a refinement step when there are already some lexical candidates to rerank.

### CODE REVIEW

**Comment by Mbeaulne** [comment]
* **#2462** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2462?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2461** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2461?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2459** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2459?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2458** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2458?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2457** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2457?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2456** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2456?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2454** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2454?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2453** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2453?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2452** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2452?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2451** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2451?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2450** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2450?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2449** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2449?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2448** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2448?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2447** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2447?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2444** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2444?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2444?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#2433** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2433?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2432** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2432?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2431** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2431?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2430** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2430?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2429** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2429?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2428** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2428?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2427** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2427?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2426** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2426?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2425** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2425?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2424** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2424?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2423** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2423?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/routes/v2/pages/Editor/components/componentSearchV2Logic.ts [hunk 1]
```diff
@@ -345,16 +345,19 @@ export function buildAiCandidateMatches(
 ): LexicalMatch[] {
   if (trimmedQuery.length === 0) return [];
 
+  const lexicalMatches = lexicalSearch(index, trimmedQuery, {
+    limit: AI_LEXICAL_CANDIDATE_LIMIT,
+    minLength: 1,
+  });
+  if (lexicalMatches.length === 0) return [];
```
Comment (camielvs) [inline src/routes/v2/pages/Editor/components/componentSearchV2Logic.ts:365]: **This gate may partially defeat the client-side embeddings added in #2432.** Semantic embeddings exist precisely to rescue matches that *lexical* scoring misses — queries with no literal token overlap. But the embedding pass runs only inside `startRerank`/`startAiSearch`, reachable only via the AI button, which is now disabled whenever lexical finds nothing (`canRerank` requires `aiCandidateMatches.length > 0`). So a purely semantic query (e.g. "shrink my tabular data" with no literal overlap to `pandas_downsample`) gets **no AI and no embedding rescue** — exactly the case embeddings were meant to cover.

If semantic-only retrieval is a goal, consider gating on "lexical **or** embedding candidates exist" rather than lexical alone. Worth confirming the intended interaction with #2432.

File: src/routes/Dashboard/DashboardComponentsV2View.tsx [hunk 2]
```diff
@@ -784,15 +767,7 @@ export const DashboardComponentsV2View = () => {
...
-    if (broadLexicalMatches.length > 0) return broadLexicalMatches;
-
-    // If literal matching found nothing, AI search can still judge a bounded
-    // browse pool. Sample evenly across the (alphabetically sorted) library so
-    // the model sees a representative spread instead of only names that sort
-    // early — matters for libraries larger than AI_CANDIDATE_LIMIT.
-    return sampleEvenly(sortedIndex, AI_CANDIDATE_LIMIT).map(
-      indexEntryToLexicalMatch,
-    );
+    return broadLexicalMatches;
```
Comment (camielvs) [inline src/routes/Dashboard/DashboardComponentsV2View.tsx:765]: **This equalizes only the *zero-hit* case; the non-zero case now diverges more between the two surfaces.** The PR says AI search is "only available as a refinement step when there are already some lexical candidates to rerank" — true here on the Dashboard (pool = top-80 lexical only). But the Editor's `buildAiCandidateMatches` still appends `buildSourceDiverseBrowseMatches(index)` + an even whole-index `sampleEvenly` fill *after* the lexical tier, so once there's **one** lexical hit it still sends a largely query-independent 80-candidate pool to the LLM, while the Dashboard sends pure lexical.

So the surfaces now behave the same on no-hit queries but noticeably differently on hit queries. If the intent is "refinement of lexical candidates," the Editor's browse-padding arguably should be removed too (ties into the #2429 query-independent-browse and #2430 surface-divergence threads). The docblock above the Editor function — "rescue matches that literal scoring missed" — is also now stale since it only runs when lexical already has ≥1 hit.

File: src/routes/v2/pages/Editor/components/componentSearchV2Logic.test.ts [hunk 3]
```diff
@@ -273,11 +273,8 @@ describe("buildLexicalMatches / buildAiCandidateMatches", () => {
     expect(buildDeepAiCandidateMatches(index, "")).toEqual([]);
   });
 
-  it("falls back to a browse pool when literal search finds nothing", () => {
-    // A query with no token overlap returns no lexical hits, so the AI
-    // candidate pool falls back to the (alphabetical) browse slice.
-    const candidates = buildAiCandidateMatches(index, "qqzznomatch");
-    expect(candidates.map((m) => m.digest)).toEqual(["alpha", "zebra"]);
+  it("returns no AI candidates when literal search finds nothing", () => {
```
Comment (camielvs) [inline src/routes/v2/pages/Editor/components/componentSearchV2Logic.test.ts:279]: Two coverage gaps worth a line each: (1) no test asserts the Editor still builds a *non-empty* padded pool when there **is** ≥1 lexical hit — the surface asymmetry noted on the Dashboard side is currently untested; (2) no test covers the lexical-empty + embeddings-available interaction, which is where the embeddings-gate concern bites.

File: src/routes/v2/pages/Editor/components/componentSearchV2Logic.ts (hunk 1 above)
Comment (Mbeaulne) [reply src/routes/v2/pages/Editor/components/componentSearchV2Logic.ts:365]: > 🤖 This is an AI-generated response.

Fixed — the no-literal-match gate now still allows smart AI search when client-side embeddings are configured, so embeddings can rescue semantic matches without falling back to query-independent browse candidates.

File: src/routes/Dashboard/DashboardComponentsV2View.tsx (hunk 2 above)
Comment (Mbeaulne) [reply src/routes/Dashboard/DashboardComponentsV2View.tsx:765]: > 🤖 This is an AI-generated response.

Fixed — dashboard smart AI search now mirrors the editor behavior more closely: it can use embeddings when lexical candidates are empty and keeps the smart rerank candidate/display limit bounded to the visible result limit.

File: src/routes/v2/pages/Editor/components/componentSearchV2Logic.test.ts (hunk 3 above)
Comment (Mbeaulne) [reply src/routes/v2/pages/Editor/components/componentSearchV2Logic.test.ts:279]: > 🤖 This is an AI-generated response.

Fixed — coverage now asserts non-empty source-diverse editor candidates when lexical search has hits and dashboard AI search availability when embeddings can rescue a zero-literal-match query.

File: src/routes/v2/pages/Editor/hooks/useComponentSearchV2State.ts [hunk 4]
```diff
@@ -295,12 +299,17 @@ export function useComponentSearchV2State(
       limit: number;
     },
   ) => {
-    if (!trimmedQuery || matches.length === 0 || !isConfigured) return;
+    if (!trimmedQuery || !isConfigured) return;
 
     const effectiveLimit =
```
Comment (maxy-shpfy) [inline src/routes/v2/pages/Editor/hooks/useComponentSearchV2State.ts]: > 🤖 This is an AI-generated code review comment.

This PR adds more branching around embedding-backed smart rerank directly inside `useComponentSearchV2State`. FTA now reports cyclomatic complexity 43 (up from 38) and FTA score 64.75 on a 363-line hook, crossing the static-analysis strong-flag thresholds. …

Suggested fix: Move the rerank-candidate assembly (`effectiveLimit`, embedding fallback, lexical truncation, and `mergeUniqueMatches`) into a pure helper near `componentSearchV2Logic.ts`, or extract an embedding/rerank sub-hook. Kee…

**Review by maxy-shpfy — APPROVED** [review]

**Comment by Mbeaulne** [comment]
### Merge activity

* **Jun 26, 7:34 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/TangleML/tangle-ui/2444).
* **Jun 26, 8:27 PM UTC**: [Graphite](https://app.graphite.com/github/pr/TangleML/tangle-ui/2444) rebased this pull request as part of a merge.
* **Jun 26, 8:30 PM UTC**: @Mbeaulne merged this pull request with [Graphite](https://app.graphite.com/github/pr/TangleML/tangle-ui/2444).
